runtime/drivers/file/repo.go

```diff
@@ -27,41 +26,31 @@ func (c *connection) DSN() string {
 }
 
 // ListRecursive implements drivers.RepoStore.
-// This implementation has some hard-coded rules: it only returns .sql files, it searches
-// to a max-depth of 3, and it excludes some common large folders (such as node_modules).
-func (c *connection) ListRecursive(ctx context.Context, repoID string) ([]string, error) {
+func (c *connection) ListRecursive(ctx context.Context, repoID string, glob string) ([]string, error) {
 	// Check that folder hasn't been moved
 	if err := c.checkRoot(); err != nil {
 		return nil, err
 	}
 
-	var paths []string
-	cleanRoot := path.Clean(c.root)
-	rootDepth := strings.Count(cleanRoot, "/")
+	fsRoot := os.DirFS(c.root)
+	glob = path.Clean(path.Join("./", glob))
 
-	err := filepath.WalkDir(c.root, func(p string, d fs.DirEntry, err error) error {
-		// Determine whether to skip the directory
+	var paths []string
+	err := doublestar.GlobWalk(fsRoot, glob, func(p string, d fs.DirEntry) error {
+		// Don't track directories
 		if d.IsDir() {
```

Comment:
Should we retain the `excludes` behaviour? Or perhaps an `includes` might be better? One that only includes models, sources and dashboards?

Reply:
I initially kept the `excludes`, but if you do something like `**/*.{yaml,sql}`, the glob will only match files (not directories), so the callback is never called for the directories, which means we don't get an opportunity to skip the excludes.

If we want proper exclude support, we need a more advanced approach. [Here's](https://github.com/moby/moby/blob/6eab4f55faf77bb0f253553d7d6fadb2ef6f4669/pkg/archive/archive.go#L885) Docker's logic for collecting files and excluding patterns in `.dockerignore`. They leverage a library they extracted here: [moby/patternmatcher](https://github.com/moby/patternmatcher).

Reply:
Ya just saw that filter. Sounds good

Reply:
The filter is not sufficient for performance though :( If you have `node_modules`, it will still traverse that entire tree looking for `.yaml` and `.sql` files...